### PR TITLE
support project monitor

### PR DESCRIPTION
**Pull request by huangzhiran**

### CODE REVIEW

File: pkg/enums/monitor_state__generated.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+	"github.com/machinefi/w3bstream/pkg/depends/kit/enum"
+)
+
+var InvalidMonitorState = errors.New("invalid MonitorState type")
+
+func ParseMonitorStateFromString(s string) (MonitorState, error) {
+	switch s {
+	default:
+		return MONITOR_STATE_UNKNOWN, InvalidMonitorState
```
Comment (CoderZhi) [inline pkg/enums/monitor_state__generated.go:19]: I didn't notice that it is an auto generated file.
@saitofun  In this case, `Unknown` is an invalid state, thus, casting returns `MONITOR_STATE_UNKNOWN, InvalidMonitorState` contains duplicate error information, `MONITOR_STATE_UNKNOWN, nil` contains conflict information.

File: pkg/modules/monitor/monitor.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	ChainHeight *CreateChainHeightReq `json:"chainHeight,omitempty"`
+}
+
+type (
+	CreateContractlogReq = models.ContractlogInfo
+	CreateChaintxReq     = models.ChaintxInfo
+	CreateChainHeightReq = models.ChainHeightInfo
+)
+
+func CreateMonitor(ctx context.Context, project *models.Project, r *CreateMonitorReq) (interface{}, error) {
```
Comment (CoderZhi) [inline pkg/modules/monitor/monitor.go:26]: New

File: pkg/enums/monitor_state__generated.go (hunk 1 above)
Comment (saitofun) [reply pkg/enums/monitor_state__generated.go:19]: error presents `ParseXXXFromString` failed, XXX_UNKNOWN is enum's default value, XXX__

File: pkg/modules/monitor/monitor.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	ChainHeight *CreateChainHeightReq `json:"chainHeight,omitempty"`
+}
+
+type (
+	CreateContractlogReq = models.ContractlogInfo
+	CreateChaintxReq     = models.ChaintxInfo
+	CreateChainHeightReq = models.ChainHeightInfo
+)
+
+func New(ctx context.Context, project *models.Project, r *CreateMonitorReq) (interface{}, error) {
```
Comment (Liuhaai) [inline pkg/modules/monitor/monitor.go:26]: why is `interface{}` used as return value?

File: pkg/models/monitor.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	MonitorID types.SFID `db:"f_monitor_id" json:"monitorID"`
+}
+
+type MonitorData struct {
+	State enums.MonitorState `db:"f_state,default='0'"     json:"-"`
+	Data  MonitorInfo        `db:"f_data"                  json:"-"`
+}
+
+type MonitorInfo struct {
+	Type        enums.MonitorType `json:"type,omitempty"`
```
Comment (saitofun) [inline pkg/models/monitor.go:34]: no need json tags ?

File: pkg/modules/monitor/monitor.go (hunk 3 above)
Comment (huangzhiran) [reply pkg/modules/monitor/monitor.go]: fixed

File: pkg/models/monitor.go (hunk 4 above)
Comment (huangzhiran) [reply pkg/models/monitor.go:34]: need this tags for json marshal and store the marshaled string into db

File: pkg/modules/monitor/monitor.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+	m := &models.Monitor{RelMonitor: models.RelMonitor{MonitorID: id}}
+	if err := m.FetchByMonitorID(d); err != nil {
+		return status.CheckDatabaseError(err, "FetchByMonitorID")
+	}
+	if project.ProjectID != m.ProjectID {
+		l.Error(errors.New("monitor project mismatch"))
+		return status.BadRequest.StatusErr().WithDesc("monitor project mismatch")
+	}
+	m.DeletedAt = bt.Timestamp{Time: time.Now()}
```
Comment (saitofun) [inline pkg/modules/monitor/monitor.go:81]: AsTimestamp(time.Now())?

File: pkg/modules/monitor/monitor.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	switch t {
+	case enums.MONITOR_TYPE__CONTRACT_LOG:
+		err = blockchain.RemoveContractLog(ctx, project.Name, id)
+	case enums.MONITOR_TYPE__CHAIN_TX:
+		err = blockchain.RemoveChainTx(ctx, project.Name, id)
+	case enums.MONITOR_TYPE__CHAIN_HEIGHT:
+		err = blockchain.RemoveChainHeight(ctx, project.Name, id)
+	}
+	return err
+}
```
Comment (saitofun) [inline pkg/modules/monitor/monitor.go:124]: Should move these APIs to `blockchain` module?

File: pkg/modules/monitor/monitor.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	if err := m.FetchByMonitorID(d); err != nil {
+		return status.CheckDatabaseError(err, "FetchByMonitorID")
+	}
+	if project.ProjectID != m.ProjectID {
+		l.Error(errors.New("monitor project mismatch"))
+		return status.BadRequest.StatusErr().WithDesc("monitor project mismatch")
+	}
+	m.DeletedAt = bt.Timestamp{Time: time.Now()}
+	m.State = enums.MONITOR_STATE__SYNCING
+	if err := m.UpdateByMonitorID(d); err != nil {
```
Comment (saitofun) [inline pkg/modules/monitor/monitor.go:83]: use `SoftDelete...` instead

File: pkg/modules/monitor/monitor.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+		l.Error(err)
+		return status.CheckDatabaseError(err, "UpdateByMonitorID")
+	}
+
+	if err := removeBlockchain(ctx, project, id, t); err != nil && err != status.NotFound {
```
Comment (saitofun) [inline pkg/modules/monitor/monitor.go:88]: use `sqlx.DBErr(err).IsNotFound()` instead

File: pkg/modules/monitor/monitor.go (hunk 6 above)
Comment (huangzhiran) [reply pkg/modules/monitor/monitor.go:124]: this not apis, just blockchain APIs wrapper

File: pkg/modules/monitor/monitor.go (hunk 7 above)
Comment (huangzhiran) [reply pkg/modules/monitor/monitor.go:83]: change to delete

**Review by saitofun — APPROVED** [review]

**Pull request closed without merging**